File: scripts/prompt-dump.ts
#!/usr/bin/env bun
/**
 * prompt-dump.ts — read and pretty-print an agent prompt/response dump.
 *
 * The local agent can write the EXACT assembled prompt + raw model response per
 * ReAct iteration to a JSONL file when `NODESPACE_PROMPT_DUMP` is set (see
 * packages/agent/src/local_agent/prompt_dump.rs). The daemon log only records the
 * system-prompt *length* and a short preview, which is insufficient for
 * diagnosing prompt/tool-call issues. This reader renders the dump so you can see
 * verbatim what reached the model and what it produced.
 *
 * Workflow:
 *   1. Run the daemon with dumping on:
 *        NODESPACE_PROMPT_DUMP=/tmp/dump.jsonl <launch nodespaced>
 *   2. Drive a turn (e.g. `bun run chat ask "Create an invoice database"`).
 *   3. Inspect:
 *        bun run prompt:dump /tmp/dump.jsonl            # full, pretty
 *        bun run prompt:dump /tmp/dump.jsonl --prompt   # system prompts only
 *        bun run prompt:dump /tmp/dump.jsonl --tools    # tools offered only
 *        bun run prompt:dump /tmp/dump.jsonl --last     # only the most recent turn
 *        bun run prompt:dump /tmp/dump.jsonl --raw      # raw JSONL lines
 *
 * Each line in the dump is one JSON object:
 *   {"kind":"turn","iteration":N,"system_prompt":"…","messages":[…],"tools":[…]}
 *   {"kind":"response","iteration":N,"raw_response":"…","tool_calls":[…]}
 */

import { readFileSync } from "node:fs";

type TurnRecord = {
  kind: "turn";
  session_id: string;
  iteration: number;
  user_message: string;
  system_prompt: string;
  system_prompt_len: number;
  messages: Array<{ role: string; content: string }>;
  tools: Array<{ name: string; description: string; parameters?: unknown }>;
};

type ResponseRecord = {
  kind: "response";
  session_id: string;
  iteration: number;
  raw_response: string;
  raw_response_len: number;
  tool_calls: Array<{ id?: string; name?: string; arguments?: string }>;
};

type Record = TurnRecord | ResponseRecord;

const args = process.argv.slice(2);
const file = args.find((a) => !a.startsWith("--"));
const flags = new Set(args.filter((a) => a.startsWith("--")));

if (!file) {
  console.error(
    "usage: bun run prompt:dump <dump.jsonl> [--prompt|--tools|--last|--raw]",
  );
  process.exit(1);
}

let lines: string[];
try {
  lines = readFileSync(file, "utf8").trim().split("\n").filter(Boolean);
} catch (e) {
  console.error(`Cannot read ${file}: ${(e as Error).message}`);
  process.exit(1);
}

if (flags.has("--raw")) {
  for (const l of lines) console.log(l);
  process.exit(0);
}

let records: Record[] = lines.map((l) => JSON.parse(l));

// --last: keep only records from the most recent session.
if (flags.has("--last")) {
  const lastSession = [...records].reverse().find((r) => r.session_id)?.session_id;
  records = records.filter((r) => r.session_id === lastSession);
}

const rule = "=".repeat(72);

for (const r of records) {
  if (r.kind === "turn") {
    if (flags.has("--tools")) {
      if (r.iteration === 0) {
        console.log(`\n${rule}\nTOOLS OFFERED (${r.tools.length})`);
        for (const t of r.tools) {
          console.log(`  • ${t.name}: ${t.description.slice(0, 90)}`);
        }
      }
      continue;
    }
    if (flags.has("--prompt")) {
      console.log(
        `\n${rule}\nTURN iter=${r.iteration} | system_prompt_len=${r.system_prompt_len}\n${rule}`,
      );
      console.log(r.system_prompt);
      continue;
    }
    // default: full
    console.log(
      `\n${rule}\nTURN iter=${r.iteration} | user: ${r.user_message}`,
    );
    console.log(`system_prompt_len=${r.system_prompt_len}`);
    console.log("--- messages sent ---");
    for (const m of r.messages) {
      const c = m.content.length > 200 ? m.content.slice(0, 200) + "…" : m.content;
      console.log(`  [${m.role}] ${JSON.stringify(c)}`);
    }
    if (r.iteration === 0 && r.tools.length) {
      console.log(`--- tools (${r.tools.length}) ---`);
      for (const t of r.tools) console.log(`  • ${t.name}`);
    }
  } else if (r.kind === "response") {
    if (flags.has("--prompt") || flags.has("--tools")) continue;
    console.log(`--- RESPONSE iter=${r.iteration} (raw_len=${r.raw_response_len}) ---`);
    if (r.raw_response) {
      const raw =
        r.raw_response.length > 300 ? r.raw_response.slice(0, 300) + "…" : r.raw_response;
      console.log(`  raw: ${JSON.stringify(raw)}`);
    }
    for (const tc of r.tool_calls) {
      console.log(`  → tool_call ${tc.name}: ${(tc.arguments ?? "").slice(0, 400)}`);
    }
  }
}

File: packages/agent/src/local_agent/prompt_dump.rs
//! Dev-only prompt/response dump.
//!
//! When `NODESPACE_PROMPT_DUMP` is set to a file path, every agent turn appends
//! the *exact* assembled system prompt, the full message list sent to the model,
//! the tools offered, and the raw (pre-normalization) model response to that
//! file as line-delimited JSON. When the env var is unset this is a zero-cost
//! no-op.
//!
//! This gives a 100%-reliable view of what actually reached the model — the
//! daemon log only records the system-prompt *length* and a short preview, which
//! is insufficient for diagnosing prompt/tool-call issues. Reach for this when
//! you need to see the verbatim prompt and the raw model output.
//!
//! Usage:
//! ```sh
//! NODESPACE_PROMPT_DUMP=/tmp/dump.jsonl <run the daemon>
//! # then inspect /tmp/dump.jsonl — one JSON object per line:
//! #   {"kind":"turn","iteration":0,"system_prompt":"<full>","messages":[...],"tools":[...]}
//! #   {"kind":"response","iteration":0,"raw_response":"<full>","tool_calls":[...]}
//! ```

use std::io::Write;

/// Env var holding the dump file path. Unset → no-op.
const ENV_DUMP_PATH: &str = "NODESPACE_PROMPT_DUMP";

/// Returns the configured dump path, or `None` when dumping is disabled.
fn dump_path() -> Option<String> {
    match std::env::var(ENV_DUMP_PATH) {
        Ok(p) if !p.trim().is_empty() => Some(p),
        _ => None,
    }
}

/// Append one JSON value as a line to the dump file. Best-effort: any IO error
/// is logged at debug and otherwise ignored — dumping must never affect a turn.
fn append(value: &serde_json::Value) {
    let Some(path) = dump_path() else { return };
    let line = match serde_json::to_string(value) {
        Ok(s) => s,
        Err(e) => {
            tracing::debug!(error = %e, "prompt_dump: serialize failed");
            return;
        }
    };
    match std::fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(&path)
    {
        Ok(mut f) => {
            if let Err(e) = writeln!(f, "{line}") {
                tracing::debug!(error = %e, path = %path, "prompt_dump: write failed");
            }
        }
        Err(e) => tracing::debug!(error = %e, path = %path, "prompt_dump: open failed"),
    }
}

/// True when dumping is enabled — lets callers skip building the JSON payload.
pub fn enabled() -> bool {
    dump_path().is_some()
}

/// Dump the exact prompt sent to the model on one ReAct iteration: the system
/// prompt, the complete (untruncated) message list — which on later iterations
/// includes accumulated tool results — and the tools offered (iteration 0 only,
/// empty thereafter to avoid repeating the schema each step).
pub fn dump_turn_iteration(
    session_id: &str,
    iteration: usize,
    user_message: &str,
    system_prompt: &str,
    messages: &[serde_json::Value],
    tools: &[serde_json::Value],
) {
    if !enabled() {
        return;
    }
    append(&serde_json::json!({
        "kind": "turn",
        "session_id": session_id,
        "iteration": iteration,
        "user_message": user_message,
        "system_prompt": system_prompt,
        "system_prompt_len": system_prompt.len(),
        "messages": messages,
        "tools": tools,
    }));
}

/// Dump the raw (pre-normalization) model response and parsed tool calls for one
/// ReAct iteration.
pub fn dump_response(
    session_id: &str,
    iteration: usize,
    raw_response: &str,
    tool_calls: &[serde_json::Value],
) {
    if !enabled() {
        return;
    }
    append(&serde_json::json!({
        "kind": "response",
        "session_id": session_id,
        "iteration": iteration,
        "raw_response": raw_response,
        "raw_response_len": raw_response.len(),
        "tool_calls": tool_calls,
    }));
}
